Refactor RealmMetadata and remove TableHandle

papafe opened the pull request:
This PR refactors `RealmMetadata` to create a correspondence between the `RealmObject.Metadata` and the `TableKey`. Given that with the `TableKey` we can get a reference to the table in native, `TableHandle` has been removed too.

Fixes #2213

nirinchev commented on Realm/Realm/Handles/ObjectHandle.cs:
```diff
@@ -66,7 +66,7 @@ private static class NativeMethods
 
             [DllImport(InteropConfig.DLL_NAME, EntryPoint = "object_get_backlinks_for_type", CallingConvention = CallingConvention.Cdecl)]
-            public static extern IntPtr get_backlinks_for_type(ObjectHandle objectHandle, TableHandle source_table, IntPtr source_property_index, out NativeException nativeException);
+            public static extern IntPtr get_backlinks_for_type(ObjectHandle objectHandle, TableKey tableKey, IntPtr source_property_index, out NativeException nativeException);
```
```suggestion
            public static extern IntPtr get_backlinks_for_type(ObjectHandle objectHandle, TableKey table_key, IntPtr source_property_index, out NativeException nativeException);
```

nirinchev commented on Realm/Realm/Handles/SharedRealmHandle.cs:
```diff
@@ -100,8 +100,8 @@ private static class NativeMethods
-            [DllImport(InteropConfig.DLL_NAME, EntryPoint = "shared_realm_get_table", CallingConvention = CallingConvention.Cdecl)]
-            public static extern IntPtr get_table(SharedRealmHandle sharedRealm, [MarshalAs(UnmanagedType.LPWStr)] string tableName, IntPtr tableNameLength, out NativeException ex);
+            [DllImport(InteropConfig.DLL_NAME, EntryPoint = "shared_realm_get_table_key", CallingConvention = CallingConvention.Cdecl)]
+            public static extern void get_table_key(SharedRealmHandle sharedRealm, [MarshalAs(UnmanagedType.LPWStr)] string tableName, IntPtr tableNameLength, out TableKey tableKey, out NativeException ex);
```
Try typedefing `realm_class_key` similar to what the C API does: https://github.com/realm/realm-core/pull/3955/files#diff-ac5ca02b88ba9910d618bac409ad7a1e3b281cdaf9f64840469c21870b7114a1R74

and have it as a return value.

nirinchev commented on Realm/Realm/Handles/SharedRealmHandle.cs:
```diff
@@ -133,10 +133,10 @@ private static class NativeMethods
 
             [DllImport(InteropConfig.DLL_NAME, EntryPoint = "shared_realm_create_object", CallingConvention = CallingConvention.Cdecl)]
-            public static extern IntPtr create_object(SharedRealmHandle sharedRealm, TableHandle table, out NativeException ex);
+            public static extern IntPtr create_object(SharedRealmHandle sharedRealm, TableKey tableKey, out NativeException ex);
```
snake_case for argument names

nirinchev commented on Realm/Realm/Native/TableKey.cs:
```diff
@@ -22,27 +22,32 @@
+        private uint value;
 
-        public bool Equals(ObjectKey other) => value.Equals(other.value);
+        public uint Value => value;
```
Not used - let's remove

nirinchev commented on Realm/Realm/Realm.cs:
```diff
@@ -1364,6 +1359,46 @@ internal void ExecuteOutsideTransaction(Action action)
+            private readonly Dictionary<string, RealmObjectBase.Metadata> stringToRealmObjectMetadataDict;
+            private readonly Dictionary<TableKey, RealmObjectBase.Metadata> tableKeyToRealmObjectMetadataDict;
+
+            public IEnumerable<RealmObjectBase.Metadata> Values => stringToRealmObjectMetadataDict.Values;
```
TODO: check where/why this is used.

papafe replied:
It is used in `realm.RemoveAll()`. It makes sense that it is used there, but probably would be more appropriate to transform the property in a method called `GetAllRealmObjectsMetadata()` or similar

nirinchev replied:
It's probably best to just implement `RemoveAll` natively. It's obviously not a hot path, but it may have peculiar side effects - e.g. if you open the Realm file with a subset of its schema, you likely expect `RemoveAll` to leave the file empty, but it'll remove only some data.

You can either leave it as is and open an issue to refactor `RemoveAll` or go ahead and fix it in this PR 😛

papafe replied:
I have added a new issue (#2233). Too much stuff already in this PR :)

nirinchev commented on Realm/Realm/Realm.cs:
```diff
@@ -1364,6 +1359,46 @@ internal void ExecuteOutsideTransaction(Action action)
+                }
+            }
+
+            public bool TryGetValue(string objectType, out RealmObjectBase.Metadata metadata)
+            {
+                return stringToRealmObjectMetadataDict.TryGetValue(objectType, out metadata);
+            }
+
+            public bool TryGetValue(TableKey tablekey, out RealmObjectBase.Metadata metadata)
+            {
+                return tableKeyToRealmObjectMetadataDict.TryGetValue(tablekey, out metadata);
+            }
```
These should be expression-bodied

nirinchev replied:
TODO: research the difference between expression bodied methods and regular ones.

nirinchev commented on Realm/Realm/Realm.cs:
```diff
@@ -1364,6 +1359,46 @@ internal void ExecuteOutsideTransaction(Action action)
+                return tableKeyToRealmObjectMetadataDict.TryGetValue(tablekey, out metadata);
+            }
+
+            public RealmObjectBase.Metadata this[string objectType]
+            {
+                get => stringToRealmObjectMetadataDict[objectType];
+            }
+
+            public RealmObjectBase.Metadata this[TableKey tablekey]
+            {
+                get => tableKeyToRealmObjectMetadataDict[tablekey];
+            }
```
TODO: check if these are safely used.

nirinchev commented on wrappers/src/object_cs.cpp:
```diff
@@ -154,11 +154,13 @@ extern "C" {
         return handle_errors(ex, [&] {
             verify_can_get(object);
 
+            auto source_table = object.realm()->read_group().get_table(tableKey);
```
maybe `const` here?

nirinchev commented on wrappers/src/shared_realm_cs.cpp:
```diff
@@ -247,11 +247,12 @@ REALM_EXPORT void shared_realm_close_realm(SharedRealm& realm, NativeException::
         Utf16StringAccessor object_type(object_type_buf, object_type_len);
-        return new TableRef(ObjectStore::table_for_object_type(realm->read_group(), object_type));
+        tableKey = ObjectStore::table_for_object_type(realm->read_group(), object_type)->get_key();
+        return;
```
unneeded return here

nirinchev commented on wrappers/src/shared_realm_cs.cpp:
```diff
@@ -354,20 +355,23 @@ REALM_EXPORT void shared_realm_write_copy(SharedRealm* realm, uint16_t* path, si
         realm->verify_in_write();
         realm->read_group();
+
+        auto table = realm->read_group().get_table(tableKey);
```
Create an inlined helper method like `get_table(SharedRealm&, TableKey)`

nirinchev commented on wrappers/src/shared_realm_cs.cpp:
```diff
@@ -444,4 +448,45 @@ REALM_EXPORT SharedRealm* shared_realm_freeze(const SharedRealm& realm, NativeEx
+        realm->verify_thread();
+
+        auto table = realm->read_group().get_table(tableKey);
+        const std::string object_name(ObjectStore::object_type_for_table_name(table->get_name()));
+        auto& object_schema = *realm->schema().find(object_name);
```
Add `find(TableKey)` in OS's Schema class and use that instead.

papafe replied:
I have just tried to implement this functionality as we discussed with @fealebenpae, by creating a map of `TableKey` to `ObjectSchema` when the `Schema` is created. Unfortunately the `table_key` field of `ObjectSchema` is not assigned yet in the constructor of `Schema`, and so we cannot create the map at that time. 
Given this, probably the fastest option would be to keep the code as it was as two operations are performed:

- `ObjectStore::object_type_for_table_name(table->get_name())` - O(1) time
- `realm->schema().find(object_name)` - O(log n) time (it uses binary search inside).

In alternative we can still implement find(TableKey) method, in two ways:

- We do a linear search - O(n) time
- We build the map the first time the method is called, and then use the map the following times - O(1) time + build time on first call

nirinchev replied:
The big-O estimates for the string case appear to ignore the string operations themselves. Not a C++ expert, but I don't believe the language uses any sort of pooling, which means that constructing the string itself and comparing it is a O(m) operation, which is exacerbated by the fact all table names start with `class_`.

Perhaps ping @fealebenpae for suggestions as that's ultimately code his team will review and maintain 😄

papafe replied:
As discussed with @nirinchev, we have decided to just use the linear search for now, then later we could eventually improve it

nirinchev commented on wrappers/src/shared_realm_cs.cpp:
```diff
@@ -444,4 +448,45 @@ REALM_EXPORT SharedRealm* shared_realm_freeze(const SharedRealm& realm, NativeEx
+        const std::string object_name(ObjectStore::object_type_for_table_name(table->get_name()));
+        auto& object_schema = *realm->schema().find(object_name);
+        if (object_schema.primary_key.empty()) {
+            const std::string name(table->get_name());
```
Pass `ObjectStore::object_type_for_table_name(table->get_name())`

nirinchev commented on wrappers/src/shared_realm_cs.cpp:
```diff
@@ -444,4 +448,45 @@ REALM_EXPORT SharedRealm* shared_realm_freeze(const SharedRealm& realm, NativeEx
+    return handle_errors(ex, [&]() {
+        realm->verify_thread();
+
+        auto table = realm->read_group().get_table(tableKey);
```
const

nirinchev commented on wrappers/src/table_cs.cpp:
```diff
@@ -51,57 +51,4 @@ REALM_EXPORT Object* table_add_empty_object(TableRef& table, SharedRealm& realm,
     });
 }
 
-REALM_EXPORT Results* table_create_results(TableRef& table, SharedRealm& realm, NativeException::Marshallable& ex)
```
This file should be deleted.

nirinchev commented on wrappers/src/object_cs.cpp:
```diff
@@ -154,11 +144,13 @@ extern "C" {
 
+            const TableRef source_table = get_table(object.realm(), table_key);
+
             const ObjectSchema& source_object_schema = *object.realm()->schema().find(ObjectStore::object_type_for_table_name(source_table->get_name()));
```
This can use the new `find(TableKey)` method.

nirinchev commented on wrappers/src/shared_realm_cs.cpp:
```diff
@@ -354,20 +356,22 @@ REALM_EXPORT void shared_realm_write_copy(SharedRealm* realm, uint16_t* path, si
+
+        const TableRef table = get_table(realm, table_key);
+
         const StringData object_name(ObjectStore::object_type_for_table_name(table->get_name()));
         const ObjectSchema& object_schema = *realm->schema().find(object_name);
```
`find(table_key)`

nirinchev commented on wrappers/src/sort_descriptor_cs.cpp:
```diff
@@ -34,13 +35,13 @@ REALM_EXPORT void sort_descriptor_destroy(DescriptorOrdering* descriptor)
         column_keys.reserve(properties_count);
 
-        const std::string object_name(ObjectStore::object_type_for_table_name(table->get_name()));
+        const std::string object_name(ObjectStore::object_type_for_table_name(get_table(realm, table_key)->get_name()));
         const std::vector<Property>* properties = &realm->schema().find(object_name)->persisted_properties;
```
This can also use `find(TableKey)`.

nirinchev reviewed (commented):
Mostly minor simplification suggestions from me, otherwise should be good to go once the Core PR is merged.

nirinchev commented on Realm/Realm/Native/TableKey.cs:
```diff
@@ -22,27 +22,18 @@
-                    return false;
-            }
-        }
+        public override bool Equals(object obj) => obj is TableKey other && value.Equals(other.value);
```
```suggestion
        public override bool Equals(object obj) => obj is TableKey other && Equals(other);
```

nirinchev reviewed (approved):
Looks good but probably should be rebased on top of master to get the build fixed.

LaPeste commented on wrappers/src/shared_realm_cs.cpp:
```diff
@@ -362,22 +364,23 @@ REALM_EXPORT void shared_realm_write_copy(SharedRealm* realm, uint16_t* path, si
-        const ObjectSchema& object_schema = *realm->schema().find(object_name);
+
+        const TableRef table = get_table(realm, table_key);
+
```
Are all these new lines needed?

LaPeste commented on wrappers/src/shared_realm_cs.cpp:
```diff
@@ -452,4 +455,44 @@ REALM_EXPORT SharedRealm* shared_realm_freeze(const SharedRealm& realm, NativeEx
+        const ColKey column_key = object_schema.primary_key_property()->column_key;
+        const ObjKey obj_key = table->find_first(column_key, from_capi(primitive));
+        if (!obj_key)
+            return nullptr;
```
Maybe this could be wrapped by curly brackets? I generally think it avoids issue.

LaPeste reviewed (changes requested):
Beside the 2 minuscule comments I had, one more. ObjectKey is renamed to TableKey and you changed the year on the copyright. I'm not sure what the policy is but I assume all the other files you modified need to be changed, as well? If not, what's the policy about copyrights?

Anyway, beside the very small things mentioned above, PR looks good to me.

papafe commented:
@LaPeste agree on both points. Regarding ObjectKey renamed to TableKey... it's just git that thinks that I renamed it. I actually deleted ObjectKey and added TableKey, but because they are similar git thinks I have renamed it, that's why the copyright difference.

papafe commented:
This PR is considered mergeable, once the core changes have been approved (https://github.com/realm/realm-core/pull/4413) and merged to the v11 branch

nirinchev commented:
The test failures are caused by a change in the Core query syntax. They'll be fixed once https://github.com/realm/realm-dotnet/pull/2214 is merged.

nirinchev commented:
Failing tests seem to be my fault - the AOT dynamic runtime doesn't seem to like my code.
